sources/core.ts

```diff
@@ -956,41 +970,37 @@ export class CommandBuilder<Context> {
+    registerDynamic(machine, node, [`isBatchOption`, this.allOptionNames], node, [`pushBatch`, this.allOptionNames]);
     registerDynamic(machine, node, [`isBoundOption`, this.allOptionNames, this.options], node, `pushBound`);
-    registerDynamic(machine, node, [`isUnsupportedOption`, this.allOptionNames], NODE_ERRORED, [`setError`, `Unsupported option name`]);
-    registerDynamic(machine, node, [`isInvalidOption`], NODE_ERRORED, [`setError`, `Invalid option name`]);
+    registerDynamic(machine, node, [`isUnsupportedOption`, this.allOptionNames], NodeType.ErrorNode, [`setError`, `Unsupported option name`]);
+    registerDynamic(machine, node, [`isInvalidOption`], NodeType.ErrorNode, [`setError`, `Invalid option name`]);
 
     for (const option of this.options) {
-      const longestName = option.names.reduce((longestName, name) => {
-        return name.length > longestName.length ? name : longestName;
-      }, ``);
-
       if (option.arity === 0) {
-        for (const name of option.names) {
-          registerDynamic(machine, node, [`isOption`, name, option.hidden || name !== longestName], node, `pushTrue`);
+        for (const name of option.aliases) {
+          registerDynamic(machine, node, [`isOption`, name], node, [`pushTrue`, option.name]);
 
           if (name.startsWith(`--`) && !name.startsWith(`--no-`)) {
-            registerDynamic(machine, node, [`isNegatedOption`, name], node, [`pushFalse`, name]);
+            registerDynamic(machine, node, [`isNegatedOption`, name], node, [`pushFalse`, option.name]);
           }
         }
       } else {
         // We inject a new node at the end of the state machine
         let lastNode = injectNode(machine, makeNode());
 
         // We register transitions from the starting node to this new node
-        for (const name of option.names)
-          registerDynamic(machine, node, [`isOption`, name, option.hidden || name !== longestName], lastNode, `pushUndefined`);
-
+        for (const name of option.aliases)
+          registerDynamic(machine, node, [`isOption`, name], lastNode, [`pushUndefined`, option.name]);
 
         // For each argument, we inject a new node at the end and we
         // register a transition from the current node to this new node
         for (let t = 0; t < option.arity; ++t) {
           const nextNode = injectNode(machine, makeNode());
 
-          // We can provide better errors when another option or END_OF_INPUT is encountered
-          registerStatic(machine, lastNode, END_OF_INPUT, NODE_ERRORED, `setOptionArityError`);
-          registerDynamic(machine, lastNode, `isOptionLike`, NODE_ERRORED, `setOptionArityError`);
+          // We can provide better errors when another option or EndOfInput is encountered
+          registerStatic(machine, lastNode, SpecialToken.EndOfInput, NodeType.ErrorNode, `setOptionArityError`);
+          registerStatic(machine, lastNode, SpecialToken.EndOfPartialInput, NodeType.ErrorNode, `setOptionArityError`);
```

Comment:
Shouldn't an option without a value work in partial mode? (Can you please add a test for it too?)

```suggestion
          registerStatic(machine, lastNode, SpecialToken.EndOfPartialInput, NodeType.SuccessNode, [`setSelectedIndex`, this.cliIndex]);
```

Reply:
No, for now I prefer `partial` to only care about the positional arguments. Reason for that is that in the use case I have, it'd likely be an error to display `foo --bar` without its option if it does have one. Reporting an error is a good way to spot anomalies.

Perhaps in the future it should be an enum with various levels of "partialness", but for now it should be good enough.